src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java

```diff
@@ -0,0 +1,82 @@
+import com.atlassian.bitbucket.jenkins.internal.client.exception.BitbucketClientException;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketPluginConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.credentials.GlobalCredentialsProvider;
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketCDCapabilities;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.cloudbees.plugins.credentials.Credentials;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Singleton
+public class DeploymentPosterImpl implements DeploymentPoster {
+
+    protected static final Logger LOGGER = Logger.getLogger(DeploymentPosterImpl.class.getName());
+
+    @Inject
+    private BitbucketClientFactoryProvider bitbucketClientFactoryProvider;
+    @Inject
+    private JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+    @Inject
+    private BitbucketPluginConfiguration pluginConfiguration;
+
+    @Override
+    public void postDeployment(String bbsServerId, String projectKey, String repositorySlug, String revisionSha,
+                               BitbucketDeployment deployment, Run<?, ?> run,
+                               TaskListener taskListener) {
+        Optional<BitbucketServerConfiguration> maybeServer = pluginConfiguration.getServerById(bbsServerId);
+        if (!maybeServer.isPresent()) {
+            taskListener.error(String.format("Could not send deployment notification to Bitbucket Server: Unknown serverId %s", bbsServerId));
+            return;
+        }
+
+        BitbucketServerConfiguration server = maybeServer.get();
+        BitbucketClientFactory clientFactory = getClientFactory(run, server);
+        BitbucketCDCapabilities cdCapabilities = clientFactory.getCapabilityClient().getCDCapabilities();
+        if (!cdCapabilities.supportsDeployments()) {
+            // Bitbucket doesn't have deployments
+            taskListener.error("Could not send deployment notification to Bitbucket Server: The Bitbucket version does not support deployments");
+            return;
+        }
+
+        taskListener.getLogger().printf("Sending notification of %s to %s on commit %s%n",
+                deployment.getState().name(), server.getServerName(), revisionSha);
+        try {
+            clientFactory.getProjectClient(projectKey)
+                    .getRepositoryClient(repositorySlug)
+                    .getDeploymentClient(revisionSha)
+                    .post(deployment);
+            taskListener.getLogger().printf("Successfully sent notification of %s deployment to %s on commit %s%n",
```

Comment:
I know IntelliJ suggests this for you, but I find it harder to read when you alternate between `printf` and `String. format`. My preference would be to pick one and stick with it. It is a personal preference though, so make your own call on it.